clippy_lints/src/transmute/missing_transmute_annotations.rs

```diff
@@ -0,0 +1,90 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use rustc_errors::Applicability;
+use rustc_hir::{GenericArg, HirId, ItemKind, Local, Node, Path, TyKind};
+use rustc_lint::LateContext;
+use rustc_middle::lint::in_external_macro;
+use rustc_middle::ty::Ty;
+
+use crate::transmute::MISSING_TRANSMUTE_ANNOTATIONS;
+
+fn get_parent_local_binding_ty<'tcx>(cx: &LateContext<'tcx>, expr_hir_id: HirId) -> Option<Local<'tcx>> {
+    let mut parent_iter = cx.tcx.hir().parent_iter(expr_hir_id);
+    if let Some((_, node)) = parent_iter.next() {
+        match node {
+            Node::Local(local) => Some(*local),
+            Node::Block(_) => {
+                if let Some((parent_hir_id, Node::Expr(expr))) = parent_iter.next()
+                    && matches!(expr.kind, rustc_hir::ExprKind::Block(_, _))
+                {
+                    get_parent_local_binding_ty(cx, parent_hir_id)
+                } else {
+                    None
+                }
+            },
+            _ => None,
+        }
+    } else {
+        None
+    }
+}
+
+fn is_function_block(cx: &LateContext<'_>, expr_hir_id: HirId) -> bool {
+    for (_, node) in cx.tcx.hir().parent_iter(expr_hir_id) {
+        if let Node::Item(item) = node
+            && let ItemKind::Fn(_, _, body_id) = item.kind
```

Comment:
`ItemKind::Fn` only applies to free standing functions. This wouldn't work for functions in `impl`s

Reply:
This would probably be easier and possibly more accurate with `enclosing_body_owner`. Would also allow us to catch `const` items like in the previous comment

Reply:
Noted!

Reply:
More like `body_owner` from what I can see from the docs (better to get a `HIrId` rather than a `DefId`).